Source/WebKit/UIProcess/win/AutomationClientWin.cpp

```diff
@@ -0,0 +1,111 @@
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "AutomationClientWin.h"
+
+#if ENABLE(REMOTE_INSPECTOR)
+#include "WebAutomationSession.h"
+#include "WebPageProxy.h"
+#include <wtf/RunLoop.h>
+#endif
+
+namespace WebKit {
+
+#if ENABLE(REMOTE_INSPECTOR)
+
+// AutomationSessionClient
+AutomationSessionClient::AutomationSessionClient(const String& sessionIdentifier)
+{
+    m_sessionIdentifier = sessionIdentifier;
+}
+
+void AutomationSessionClient::requestNewPageWithOptions(WebKit::WebAutomationSession& session, API::AutomationSessionBrowsingContextOptions, CompletionHandler<void(WebKit::WebPageProxy*)>&& completionHandler)
+{
+    bool found = false;
+    auto processPool = session.protectedProcessPool();
+    if (processPool && processPool->processes().size()) {
+        // Set controlledByAutomation flag to all pages in all processes
+        processPool->setPagesControlledByAutomation(true);
+        // The default page for automation target is the first page of the first process
+        auto& processProxy = processPool->processes()[0];
+        if (processProxy->pageCount()) {
+            found = true;
+            completionHandler(processProxy->pages()[0].ptr());
+        }
+    }
+    if (!found)
+        completionHandler(nullptr);
+}
+
+void AutomationSessionClient::didDisconnectFromRemote(WebKit::WebAutomationSession& session)
+{
+    session.setClient(nullptr);
+
+    RunLoop::main().dispatch([&session] {
+        auto processPool = session.protectedProcessPool();
+        if (processPool) {
+            processPool->setAutomationSession(nullptr);
+            processPool->setPagesControlledByAutomation(false);
+        }
+    });
+}
+
+// AutomationClient
+AutomationClient::AutomationClient(WebProcessPool& processPool)
+    : m_processPool(processPool)
+{
+    Inspector::RemoteInspector::singleton().setClient(this);
+}
+
+AutomationClient::~AutomationClient()
+{
+    Inspector::RemoteInspector::singleton().setClient(nullptr);
+}
+
+void AutomationClient::requestAutomationSession(const String& sessionIdentifier, const Inspector::RemoteInspector::Client::SessionCapabilities&)
+{
+    ASSERT(isMainRunLoop());
+
+    auto session = adoptRef(new WebAutomationSession());
+    session->setSessionIdentifier(sessionIdentifier);
+    session->setClient(WTF::makeUnique<AutomationSessionClient>(sessionIdentifier));
+    m_processPool->setAutomationSession(WTFMove(session));
+}
+
+void AutomationClient::closeAutomationSession()
+{
+    RunLoop::main().dispatch([this] {
+        if (!m_processPool->automationSession())
```

Comment:
I think you have to check the WeakPtr before using it.
	   
~~~
  if (!m_processPool)
    return;
~~~

Reply:
You have to keep a RefPtr of m_processPool to call its methods.
https://github.com/WebKit/WebKit/wiki/Smart-Pointer-Usage-Guidelines
> RefPtr processPool = m_processPool;
> if (!processPool)
>     return;

Reply:
I think it might be a good idea to add a new getter `protectedProcessPool()`.